BREAKING: Move cli resources to sensu::cli class

treydock opened the pull request:
# Pull Request Checklist

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Move cli resources to sensu::cli class
Manage sensu user/group so to support sensu::cli on host not running agent or backend

* Move sensu::backend::cli_package_name to sensu::cli::package_name
* Move sensu::backend::url_host to sensu::api_host
* Move sensu::backend::url_port to sensu::api_port
* Move sensu::backend::password to sensu::password
* Move sensu::backend::old_password to sensu::old_password
* Move sensu::backend::agent_password to sensu::agent_password
* Move sensu::backend::agent_old_password to sensu::agent_old_password
* Move sensu::backend::sensuctl_chunk_size to sensu::cli::sensuctl_chunk_size
* Replace `sensu_config` type with `sensu::cli::config_format` and `sensu::cli::config_namespace` parameters.

Always manage `sensu:agent` user to ensure password set
Set `password` in agent.yml


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
This will allow non-backends to have sensuctl setup which is supported by Sensu Go. This also would make it easier to have a Bolt task that can use sensuctl to populate Bolt Inventory v2.

The management of sensu user/group was necessary to support using sensu::ctl class on a host that doesn't include sensu::agent or sensu::backend as the agent and backend packages had previously been what adds the user/group.

treydock commented:
Marked WIP as needs support for Windows and possibly addressing #1163 as it was the acceptance tests for sensu::cli where I ran into #1163.

ghoneycutt commented on manifests/cli.pp:
```diff
@@ -0,0 +1,72 @@
+    bootstrap_password => 'P@ssw0rd!',
+    trusted_ca_file    => $trusted_ca_file,
+  }
+
```
no empty lines between `}`

treydock replied:
Fixed.

ghoneycutt commented on manifests/cli.pp:
```diff
@@ -0,0 +1,72 @@
+    username           => 'admin',
+    password           => $password,
+    bootstrap          => $bootstrap,
+    bootstrap_password => 'P@ssw0rd!',
```
should be a param

treydock replied:
I would disagree, changing the value would not work. This value is the default value from Sensu Go and anything but this value will break the module on fresh installs. We have to bootstrap using default password.

treydock replied:
To avoid confusion I will hardcode the value into the Ruby code where it's used. If the value is changed by anyone using this module it will break that functionality on fresh installs so probably better to avoid confusion and just not expose it at all as parameter. The only time we'd have to change the value is if upstream Sensu Go starts using a different default password.

treydock replied:
Bootstrap parameters removed.

ghoneycutt commented on README.md:
````diff
@@ -130,6 +131,19 @@ associated to `linux` and `apache-servers` subscriptions.
+```
+
+**NOTE**: The `sensu::backend` class calls the `sensu::cli` class so it is only necessary to directly call the `sensu::cli` class on hosts not using the `sensu::backend` class.
+
````
We should further explain what bootstrapping entrails and why it is needed.

treydock replied:
The bootstrapping only happens on fresh install of backend. It's something we abstract away and not something I would expect anyone to need to worry about since we just recommend not using default password. The bootstrapping is not needed for just using cli class like this section documents. This is kinda of like a private type functionality. I can document but if module behaves correctly people should be unaware of bootstrap step.

treydock replied:
Also I hope to rework some the logic to avoid need to pass bootstrap parameters. The current process is semi fragile per #1163. Somewhere still have to hardcode the default bootstrap password though.

ghoneycutt replied:
This can be a commented out section in the code or the README as you see fit, though a few sentences regarding How Sensu ships with a default password and we have to go through this bootstrap process to change the password would help users understand what is happening.

treydock replied:
I improved the bootstrap process so it's entirely hidden. We document using non-default password and everything else is handled by the provider for `sensu_configure`.  See d2ef64b07805262733c87ccd2bd28de89d2c9a4f

ghoneycutt commented on manifests/cli.pp:
```diff
@@ -0,0 +1,113 @@
+# @param password
+#   Sensu backend admin password used to confiure sensuctl.
+# @param bootstrap
+#   Should sensuctl be bootstrapped. This is a private parameter used by sensu::backend class.
```
it is used in this class, too

treydock replied:
I found a way to avoid bootstrap parameters. This has been removed.

ghoneycutt commented on manifests/cli.pp:
```diff
@@ -0,0 +1,113 @@
+      path       => $sensuctl_path,
+    }
+
+    sensu_configure { 'puppet':
```
is this duplicating code in `sensu::bootstrap` ?

treydock replied:
This used to be in `sensu::backend` and was moved here and backend just class `cli` class.  I removed all bootstrapping parameters from this PR.

treydock commented:
Removed WIP. I will fix #1163 in a different pull request that will have to incorporate this PR. The changes necessary to fix #1163 are rather involved and potentially not backwards compatible as a parameter will have to get removed.

treydock commented:
I am struggling to get simple "Does file exist?" to work in appveyor. I got 2 appveyor instances stuck on RDP but didn't accept my password so have to wait till they time out. I'll cleanup windows testing once I'm able to debug why checking file existence is not working.

ghoneycutt commented on .fixtures.yml:
```diff
@@ -36,5 +36,8 @@ fixtures:
     archive:
       repo: git://github.com/voxpupuli/puppet-archive.git
       ref: 'v3.0.0'
+    windows_env:
```
Why is this needed?

treydock replied:
This is needed to set `PATH` for sensuctl on Windows since the install isn't a package, it's just a zip file.

ghoneycutt commented on appveyor.yml:
```diff
@@ -48,16 +49,19 @@ test_script:
-    - PUPPET_GEM_VERSION: '~>6.x'
-      PUPPET_REPO: puppet6
-      RUBY_VERSION: 25-x64
+#    - PUPPET_GEM_VERSION: '~>6.x'
```
why are these commented out?

treydock replied:
See my recent comment yesterday, I am debugging appeveyor failures and it's easier to iterate over changes with fewer changes running. I had to pause my work last night due to issues with RDP in appveyor builds that allows me to log into VM and see why tests are failing. I will revert these changes once I get appveyor working.

ghoneycutt commented on lib/puppet/provider/sensu_api.rb:
```diff
@@ -0,0 +1,28 @@
+    return false
+  end
+end
+
```
trailing newline

treydock replied:
Fixed.

ghoneycutt commented on lib/puppet/type/sensu_configure.rb:
```diff
@@ -35,11 +35,6 @@
     desc "Password to use with sensuctl configure"
   end
 
-  newparam(:bootstrap_password) do
```
Changing the public interface feels like a breaking change.

treydock replied:
If anyone ever changed this value then it would have resulted in errors. The default value was the only value that would work as it's the Sensu Go default admin password. Also we do not document the use of `sensu_configure` as it's kind of a private type that we use internally in the backend class and now moved to the cli class. If anyone changed this value then their install would have not worked so I think it's safe to consider this backwards compatible.

ghoneycutt commented on manifests/backend.pp:
```diff
@@ -176,6 +179,7 @@
   Stdlib::Port $postgresql_port = 5432,
   String $postgresql_dbname = 'sensu',
   Integer $postgresql_pool_size = 20,
+  Boolean $declare_cli_class = true,
```
This is a new patter, why should the user care how the class is included?

treydock replied:
Because if someone wants to do this:

```puppet
class { '::sensu::cli': ... }
class { '::sensu::backend': ... }
```

Then they will duplicate resource errors unless they set `declare_cli_class => false`.

ghoneycutt replied:
Could we use data in module to work around this?

treydock replied:
I don't know how that would work? This is something I think we just have to allow someone to turn on/off based on how they use this module.

treydock replied:
I think only way to avoid this would be to do something like `if ! defined(Class['sensu::cli'])` but I thought such order dependent hacks were discouraged in Puppet.

ghoneycutt replied:
Why do we have to use `class` at all instead of using`include`?

treydock replied:
Because then instead of this:

```
class { 'sensu::backend': url_host => 'foo' }
```

A site has to do this with just include: 

```
class { 'sensu::cli': url_host => 'foo' }
class { 'sensu::backend': url_host => 'foo' }
```

There are parameters that we pass to backend class that are also needed by CLI class and this maintains backwards compatibility for behavior too.

treydock replied:
One idea is to move `url_host` and `url_port` to `sensu` class since shared between backend agent and now CLI. This would be easier once #1159 is merged and #1159 will have to be updated once this PR is merged. If we move these and other parameters to `sensu` class we could simplify with `include` but would be breaking change. I can make that change as part of #1159

ghoneycutt replied:
I like that approach. Let's do this as part of the breaking change.

treydock replied:
So implement that breaking change in this PR or keep this current behavior and do new PR for breaking change?

treydock replied:
This was implemented.

treydock commented:
Fixed Windows tests and restored previous test matrix. This PR should be good to go once Travis tests complete.

treydock commented:
Closed #1159 as this had similar changes that made that PR not necessary.

treydock commented:
I've rebased with master and added Bolt v2 inventory task that requires the CLI changes in order to run inventory on non-backend node.

treydock commented:
Added commit to make managing agent passwords much more consistent. Updated PR description as moved `agent_password` and `agent_old_password` from `sensu::backend` class to `sensu` class. This way both backends and agents could use same value and ensure that value is propagated to `sensu:agent` user and also updated in `agent.yml`. Previous behavior was flawed and made changing the password very difficult.

treydock commented:
Added commit to consolidate `sensu_config` which was just a way to set a few `sensuctl` settings and moved those settings into `sensuctl_configure` and exposed them via `sensu::cli` class. The previous `sensu_config` would allow invalid settings and had no validation and didn't really do much that couldn't be handled by the type managing sensuctl configuration.
